### ISSUE TITLE
Testing controllers (services) with DI for FastifyInstance

### ISSUE THREAD

**Issue by unematiii** [L2jLiga/fastify-decorators#84]
Hey!

First of all, thank you for this amazing project!

I'm having a bit of trouble testing a controller which has DI for FastifyInstance, st like this:
```ts
@Inject(FastifyInstanceToken)
  private instance!: FastifyInstance;
```

However, I cant seem to get it working with `configureControllerTest`:
```
TypeError: Invalid argument provided for "MyController.instance". Expected class annotated with @Service.

      at injectProperties (lib/decorators/helpers/inject-dependencies.ts:39:19)
      at Object.createWithInjectedDependencies (lib/decorators/helpers/inject-dependencies.ts:29:5)
      at Object.0 (lib/decorators/strategies/controller-type.ts:42:36)
      at instance.register.prefix (lib/decorators/controller.ts:40:85)
      at Plugin.Object.<anonymous>.Plugin.exec (node_modules/avvio/plugin.js:131:17)
      at Boot.loadPlugin (node_modules/avvio/plugin.js:266:10)
      at Task.release (node_modules/fastq/queue.js:143:16)
      at worked (node_modules/fastq/queue.js:195:10)
      at node_modules/avvio/plugin.js:269:7
      at done (node_modules/avvio/plugin.js:201:5)
```

Also, as a side note, It might be good to mention somewhere in the docs, for whoever is using ts-jest, that they need to properly set up jest environment, i.e. the crucial part:
```js
require('reflect-metadata');
```

**Comment by L2jLiga** [comment]
Hi there, first of all thank you for your feedback!

Regarding to using `FastifyInstance` in tests it seems like you found a bug in library.
Regarding to `reflect-metadata` it seems unclear what we can do on it, I see 2 options: 1st is two describe it in documentation and 2nd is two implicitely require `reflect-metadata` inside `configureServiceTest` and `configureControllerTest`. What is you vision in this topic?

btw PRs are always welcomed :)

**Comment by unematiii** [comment]
Hey!

Regarding `reflect-metadata` issue. I think its a probably a common sense to have this set up for your tests, but then again, st that can go easily missed if you test separate parts of your application... I think mentioning it somewhere in the docs is probably sufficient (an example of jest env setup is already in the repo).

I managed to get my tests working with [these](https://gist.github.com/unematiii/0f74b6d7746ba2485dce198b50398e62) additions, but I'm not sure if this is the right way of doing it (as I'm just learning how the whole lib works...)...

Also, I was thinking, when using `configureControllerTest`, it would be nice to get a reference to the underlying controller instance itself (to spy of some methods etc...).

**Comment by L2jLiga** [comment]
- I added your patch to codebase with small additions, now it will provide fastify instance in case it was not mocked.
- With `reflect-metadata` I will think how can I mention it in documentation (as of now I think it looks quite mess).
- Returning reference to underlying controller instance would not be such easy, in general what library need to do is to store instance somewhere to be able get it in `configureControllerTest`, as of now it does not store it)

**Comment by unematiii** [comment]
@L2jLiga Thanks! Any chance you will publish this?

**Comment by L2jLiga** [comment]
Yeah, did it right now, version 3.4.1 is available on npm :)

**Comment by unematiii** [comment]
Thanks, upgraded and works nicely. Think we can close this.

**Comment by L2jLiga** [comment]
I would keep it open to not forget about documentation improvements and returning controller instance

**Comment by L2jLiga** [comment]
Added an option to access controller instance in tests, also reworked documentation:

https://github.com/L2jLiga/fastify-decorators/blob/v3/docs/Testing.md

**Comment by L2jLiga** [comment]
Released as 3.5.0
